text/0000-inline-const.md

```diff
@@ -240,37 +246,77 @@ that they are best considered as a unit, however.
 # Unresolved questions
 [unresolved-questions]: #unresolved-questions
 
+## Generic Parameters
+
+Named constants defined inside a function cannot use generic parameters that
+are in scope within that function.
+
+```rust
+fn foo<T>() {
+    const SIZE: usize = { std::mem::size_of::<T>() }; // ERROR
+}
+```
+
+If we use the same rule for inline constants, there would be a class of
+implicitly promotable expressions (e.g. `std::ptr::null::<T>()`, `T::CONST +
+T::ANOTHER_CONST`), that could not be written inside an inline constant. We
+would need to resolve this somehow before we stop promoting these expressions
+as discussed in [future possibilites].
+
+Alternatively, we could allow inline constants to refer to generic parameters.
+Associated constants already behave this way, so I think this is possible.
+However, it may result in more post-monomorphization const-eval errors.
+
 ## Naming
 
 I prefer the name inline `const`, since it signals that there is no difference
-between a named `const` and an inline one. @scottmcm prefers "`const` block",
-which is closer to the syntax and parallels the current terminology of `async`
-block and `unsafe` block. It also avoids any accidental conflation with the
-`#[inline]` attribute, which is unrelated.
+between a named `const` and an inline one.
+
+@scottmcm prefers "`const` block", which is closer to the syntax and parallels
+the current terminology of `async` block and `unsafe` block. It also avoids any
+accidental conflation with the `#[inline]` attribute, which is unrelated.
+Additionally, it doesn't extend nicely to the single-expression variant
+discussed in [future possibilities].
+
+@RalfJung prefers "anonymous `const`". @scottmcm mentioned in Zulip that this
+could be confused with the `const _: () = ...;` syntax introduced in [RFC
+2526]. The reference refers to these as "unnamed" constants.
+
+[RFC 2526]: https://github.com/rust-lang/rfcs/pull/2526
 
 ## Lints
 
 As mentioned in the reference-level specification, we need to decide whether we
 want to lint against certain types of inline `const` expressions.
 
 # Future possibilities
-[future-possibilities]: #future-possibilities
+[future possibilities]: #future-possibilities
 
 It would be possible to allow the syntax `const expr` for an inline `const` that
 consists of a single expression. This is analagous to the single expression
 variant of closures: `|| 42`. This is backwards compatible with the current proposal.
 
-This could allow us to deprecate the more esoteric classes of promotable
-expressions (e.g., `&(u32::MAX + u32::MAX)`) in favor of inline `const`
-expressions. This would have to be done at an edition boundary. We would only
-do promotion for aggregates, literals, constants and combinations thereof, and
-`#[rustc_promotable]` would be removed from the standard library.
+Eventually, I would like to try making any expression that could possibly panic
+inelgible for implicit promotion. This includes *all* `const fn` calls as well
```

Comment:
I am not sure if making panics the main motivation here is the best argument. (I agree with your goal here, I just think we can more convincingly argue for it.) As we discussed, some bit-ops cannot panic, and the functions marked `#[rustc_promotable]` also cannot panic. (You say that later but it still looks like panics are the main motivation.)

I think a good lead-in here is to state the goal of unifying implicit promotion and patterns: `&expr` is implicitly promoted iff `expr` is valid as a pattern. That is easy to remember and it means there's one special "class" of things less to learn about. (This idea is not by me; it is [by @eddyb](https://github.com/rust-lang/const-eval/issues/19#issuecomment-472799062).)

Then we can argue that s a nice side-effect, we would then also rule out panics in implicitly promoted expressions.

(Incidentally, this is also the best argument for allowing `const { ... }` in patterns that I have seen so far.)

Reply:
This thread is still open.